### ISSUE TITLE
rustc requires 7GB of memory for clean build

### ISSUE THREAD

**Issue by samuela** [rust-lang/rust#67183]

**Comment by jonas-schievink** [comment]
`-Ztime-passes` reports that borrow checking is the first pass that makes the memory usage grow a lot:

```
  time: 1.021; rss: 1383MB	misc checking 2
  time: 11.567; rss: 4937MB	MIR borrow checking
```

This *is* almost a 500000 line crate consisting almost entirely of generated code though, so rustc using lots of memory should be very much expected.

**Comment by samuela** [comment]
@jonas-schievink Very cool, I was not aware of `-Ztime-passes`! Does the borrow checker run its constraint solver on the level of entire crates or modules? Although the crate is relatively large, the borrow checking process seems like it could be done in a modular fashion.

**Comment by samuela** [comment]
I'm not able to run with that flag, however:
```
$ cargo build --jobs 1 -Ztime-passes
error: unknown `-Z` flag specified: time-passes
```

**Comment by jonas-schievink** [comment]
It should run on the function-level, but maybe it keeps around more data than necessary. Or there's a huge function in the crate that makes it choke.

**Comment by jonas-schievink** [comment]
`-Ztime-passes` is a rustc flag, you can run `cargo rustc --jobs 1 -- -Ztime-passes`

**Comment by samuela** [comment]
> It should run on the function-level, but maybe it keeps around more data than necessary. Or there's a huge function in the crate that makes it choke.

Yeah, I guess I'd really need to know which function requires this much memory. It would have to be truly huge.

**Comment by samuela** [comment]
@jonas-schievink I'm getting quite different numbers on the `MIR borrow checking` step:
```
  time: 24.250; rss: 2916MB	MIR borrow checking
```
This section looks more problematic however:
```
  time: 0.000; rss: 2030MB	metadata encoding and writing
      time: 0.082; rss: 2031MB	collecting roots
      time: 11.252; rss: 4473MB	collecting mono items
    time: 11.334; rss: 4473MB	monomorphization collection
    time: 0.043; rss: 4474MB	codegen unit partitioning
    time: 0.000; rss: 4474MB	write allocator module
    time: 0.001; rss: 2889MB	llvm function passes [1edthmlfja75g550]
    time: 0.158; rss: 2891MB	llvm module passes [1edthmlfja75g550]
    time: 4.987; rss: 3242MB	codegen passes [1edthmlfja75g550]
    time: 0.002; rss: 2195MB	llvm function passes [19d4spp43mmbwzve]
    time: 0.024; rss: 2195MB	llvm module passes [19d4spp43mmbwzve]
    time: 0.466; rss: 2190MB	codegen passes [19d4spp43mmbwzve]
```

What does `monomorphization collection` mean?

Here's a dump of the complete log: https://pastebin.com/zYHa6KNG

**Comment by jonas-schievink** [comment]
> What does `monomorphization collection` mean?

The module docs explain it better than I could: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/monomorphize/collector/index.html

**Comment by samuela** [comment]
I just upgraded to rustc 1.41.0-nightly (412f43ac5 2019-11-24) and re-ran: https://pastebin.com/7m3sSaAr. 

`time -v` reports that the max RSS is >7.5GB, but `-Ztime-passes` output maxes out at 
```
    time: 0.000; rss: 4753MB	write allocator module
```
which is ~4.6GB. So where is the mystery 2.9GB coming from?

**Comment by jonas-schievink** [comment]
`-Ztime-passes` only measures after each pass, it misses any transient allocations happening inside of them. A heap profiler such as DHAT should provide more insight.

**Comment by samuela** [comment]
@jonas-schievink Ah, I see. I'll try to get DHAT set up.

**Comment by Enselic** [comment]
Triage: Can this still be reproduced with latest stable/nightly?

**Comment by Enselic** [comment]
I tried to build 37c0a1d in https://github.com/samuela/rustybox with latest stable but got a lot of strange errors, but no RAM increase.

Does the reproducer work for anyone any longer? If not, I suggest we close this issue as obsolete.

**Comment by Enselic** [comment]
Triage: Closing as stale.
